src/context/AuthStore.tsx

```diff
@@ -0,0 +1,135 @@
+import React, { useReducer, useEffect } from 'react';
+import { AuthPayload, Mutation } from 'types';
+import { useMutation } from '@apollo/react-hooks';
+import { auth } from 'utils';
+import LOGIN_MUTATION from 'graphql/login.graphql';
+
+enum ActionTypes {
+  AUTH_INIT = 'auth/init',
+  AUTH_ERROR = 'auth/error',
+  AUTH_SUCCESS = 'auth/success',
+  AUTH_LOGOUT = 'auth/logout',
+}
+
+interface State {
+  loading: boolean;
+  error: any;
+  data: AuthPayload;
+  isAuthenticated: boolean;
+}
+
+interface Action {
+  type: string;
+  payload?: any;
+}
+
+export interface AuthContextData extends State {
+  login: (email: string, password: string) => Promise<any>;
+  logout: () => void;
+}
+
+const defaultState: State = {
+  loading: false,
+  error: null,
+  data: {} as AuthPayload,
+  isAuthenticated: false,
+};
+
+const defaultContext: AuthContextData = {
+  ...defaultState,
+  login: () => new Promise(() => null),
+  logout: () => null,
+};
+
+export const AuthContext = React.createContext(defaultContext);
+
+const authReducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case ActionTypes.AUTH_INIT:
+      return {
+        ...state,
+        loading: true,
+        error: null,
+      };
+    case ActionTypes.AUTH_SUCCESS:
+      return {
+        ...state,
+        loading: false,
+        data: { ...action.payload },
+        isAuthenticated: true,
+      };
+    case ActionTypes.AUTH_ERROR:
+      return {
+        ...state,
+        loading: false,
+        error: { ...action.payload },
+      };
+    case ActionTypes.AUTH_LOGOUT:
+      return {
+        ...state,
+        data: {},
+        isAuthenticated: false,
+      };
+    default:
+      throw new Error();
+  }
+};
+
+const AuthStore = (props: any): JSX.Element => {
+  const [state, dispatch] = useReducer(authReducer, defaultState);
+  const [loginMutation] = useMutation<Mutation>(LOGIN_MUTATION, {
```

Comment:
`useMutation<Mutation>` shouldn't be the generated types for login mutation?

Reply:
So it is:) We importing it from ```import { AuthPayload, Mutation } from 'types';``` And `types` here is generated types from `/src/types/schema-types.ts`

Reply:
Why? Generated types provide us with strickt types for all mutations

Reply:
True, I miss understood the code :P